[RNMobile] Add Contrast Checker to Text-Based Blocks

SiobhyB opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/33396

`Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4357

## Description

With this PR, a contrast checker is added to the text-based blocks in the native apps. This checker will display the following notice when users pick a custom text/background colours with poor contrast:

```
This colour combination may be hard for people to read. Try using a brighter background colour and/or a darker text colour.
```

## How has this been tested?

* Add a new text block (e.g. the paragraph block) to a new a post in the mobile app and type some text.
* Tap the block's cog/gear icon to edit its settings.
* Select a foreground and background colour that have a low contrast to one another (e.g. two really dark colours).
* The new notice should appear to inform you that the contrast is low between the two selected colours.

## Testing Notes

* If a theme _isn't_ block-based then the contrast checker will only run when both a custom text and a custom background colour are set for a block i.e. it won't run if a default colour is being used. This is because there isn't any ideal or straightforward way to determine the default colours of a theme that _isn't_ block-based. The web is able to use [getComputedStyles()](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle) but this isn't an option for native development. 
* As this change is intended to help accessibility, we should ensure that the notice itself is accessible. The contrast should meet AA standards and be readable via screen readers.
* To keep the scope of the PR narrow, the notice is only introduced to the main settings screen of blocks that have a colour panel in their settings. The notice could be added to individual screens in future iterations.

## Screenshots <!-- if applicable -->

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/145132004-c8f7a3ea-128d-4207-bf28-74f239f6a801.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/145125553-321dba53-533d-4847-9e06-e9b5773996c5.png" width="100%"> |

## Video

https://user-images.githubusercontent.com/2998162/145385504-bc1a1872-1d7b-4959-a290-abc9cbda6d38.mov

## Types of change

New feature (non-breaking change which adds functionality), with the following high-level overview of the code:

* A native version of [`components/contrast-checker`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/contrast-checker/README.md) has been created for use with the apps.
* In addition, a native version of [`hooks/color-panel`](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/colors) has been created. This was necessary in order to detect active text/background colours from the apps.
*  `components/contrast-checker` is called from `hooks/color-panel.native` in order to run the checker within blocks that call that hook.

<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Enable ContrastChecker usage on mobile

With this commit, .native files have been created for the ContrastChecker component.

The component's main index.native.js was copied over from the main web version, with web-specific pieces (e.g.<div> elements) removed so that it can be called from other mobile files without error.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/color-contrast-checker

SiobhyB committed:
Adapt ColorPanel hook for native use

This commit takes the web's ColorPanel hook and adapts it for native use.

Specifically, while the web version relies on ComputedStyles to fetch a block's background/text colour, we need to use other means for native.

Here, we're using the useGlobalStyles, which will fetch a block's background and text colour when they're explicitly set via the global style settings.

SiobhyB committed:
Add ContrastChecker to Color panel

With this commit, the ContrastChecker component (which is being passed via "children") is added to the Color panel.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/color-contrast-checker

SiobhyB committed:
Style new notice for light and dark mode

SiobhyB committed:
Detect changes to block's text and/or background color

With this commit, useEffect and useState hooks are introduced to detect changes to a block's text and/or background colors.

SiobhyB committed:
Remove redundant/unused CSS

SiobhyB commented:
👋  @iamthomasbishop, @kyleaparker, would I be able to get your feedback on the current design of the notice? I used `$gray-text-min` for light mode (the lightest shade of gray that passes AA contrast checks) and white for dark mode:

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/145132004-c8f7a3ea-128d-4207-bf28-74f239f6a801.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/145125553-321dba53-533d-4847-9e06-e9b5773996c5.png" width="100%"> | 

Looking forward to hearing any feedback!

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/color-contrast-checker

SiobhyB commented:
👋 @geriux, would you have time to review this? As noted in the description, this PR only adds the contrast checker to the main colour settings panel, as I thought it'd be best to keep the scope fairly narrow for now. It could be added to the individual colour picker screens in future iterations.

Let me know if anything is unclear in the code or description! Looking forward to hearing any feedback.

geriux commented:
> 👋 @geriux, would you have time to review this?

Hey 👋 Of course! looking forward to review/test this feature 🎉

iamthomasbishop commented:
@SiobhyB This is looking pretty solid, nice work! At first glance there are a few of things I'm seeing, for which I'll make proposals for:

- We can remove the separator between the background-color and footnote cells
- This might look a bit too strong but I'd like to try changing the text label to yellow (`$yellow-50` in light mode and `$yellow-30` in dark mode), at least to see if it feels right in context.
- In addition, would it be possible to add a `warning` icon (from [wp/icons](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library)) so that it sits aligned left and vertically-centered with the label? I'm guessing this might be tricky unless our footer cell allows it but it would be a nice visual cue.
- The last thing is that if `$yellow-` doesn't work, I think the label color you're using in dark mode is a bit too strong. I believe it should be the same color as the section header labels but I'm not 100% sure as it's using the "old" colors.

Here's what that would look like all together:

![1000](https://cldup.com/lp3Jz32VTC.png)

_As a separate note: in the long-term, it might be nice to try also adding a contrast warning on the sub-sheet where you select an individual color, but this is a great start._

Thanks for working on this and great job! It'll be a really handy feature that our users will love! 😍

geriux commented on packages/block-editor/src/components/contrast-checker/index.native.js:
```diff
@@ -0,0 +1,99 @@
+ * External dependencies
+ */
+import { Text } from 'react-native';
+import tinycolor from 'tinycolor2';
```
It looks like the [tinycolor](https://github.com/WordPress/gutenberg/pulls?q=is%3Apr+tinycolor) usage was replaced in favor of `colord`. Do you think we could use it as well for this feature?

SiobhyB replied:
Ah, thank you for spotting that! It looks like `colord` was introduced after I began work on this and I didn't notice. I updated the references in https://github.com/WordPress/gutenberg/pull/34902/commits/215f95cc9bf6d207cfe6532729b7d419fb71e3b7.

geriux commented on packages/block-editor/src/components/contrast-checker/index.native.js:
```diff
@@ -0,0 +1,99 @@
+import { speak } from '@wordpress/a11y';
+import { __ } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
+import { withPreferredColorScheme } from '@wordpress/compose';
```
Since this is a functional component we could use `usePreferredColorSchemeStyle` instead of `withPreferredColorScheme` to avoid using a HOC. Then we can just replace `getStylesFromColorScheme(...)` to `usePreferredColorSchemeStyle(...)`

SiobhyB replied:
I went ahead to implement this in https://github.com/WordPress/gutenberg/pull/34902/commits/0630f8913478ad337e724cddef592bbeaa98aece, thanks!

geriux commented on packages/block-editor/src/components/contrast-checker/style.native.scss:
```diff
@@ -0,0 +1,9 @@
+.message {
```
Can we update the names to follow the [BEM methodology](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#naming)? A few months back we started following the same guidelines as in core.

SiobhyB replied:
I did this as part of styling changes in https://github.com/WordPress/gutenberg/pull/34902/commits/c37e2dafdfc19ac68ab62b9387dd561fe3c320c0. This is my first time following that methodology, so let me know if anything looks off to you!

geriux replied:
Looks good! Thanks for the changes!

geriux commented on packages/block-editor/src/hooks/color-panel.native.js:
```diff
@@ -0,0 +1,67 @@
+};
+
+export default compose( [
+	withSelect( ( select ) => {
```
Since we have a functional component we could use the `useSelect` hook instead of wrapping the component with `withSelect`.

SiobhyB replied:
Thank you for this tip! I've gone ahead to implement this in https://github.com/WordPress/gutenberg/pull/34902/commits/822389b2549b7faf904483e204e4523be79604a7.

geriux reviewed (commented):
Hey @SiobhyB 👋 great work adding this feature! I tested different scenarios and its working correctly 🎉 . I left a couple of minor comments, let me know what you think.

By the way, it'd be great if you could update from the latest changes in `trunk` so I can also test the higlighting text feature. Thanks!

SiobhyB committed:
Update 'tinycolor' references with 'colord'

'tinycolor' usage has recently been updated in core with 'colord', so this commit reflects that update.

SiobhyB committed:
Replace 'withSelect' HOC with 'useSelect' hook

As we're using a functional component, the use of hooks are preferable to HOCs.

SiobhyB committed:
Replace 'withPreferredColorScheme' HOC with hook

As we're using a functional component, the use of hooks are preferable to HOCs. The 'withPreferredColorScheme' HOC is therefore replaced with the 'usePreferredColorSchemeStyle' as part of this commit.

SiobhyB committed:
Styling changes; warning icon and colour change

This commit includes changes to the styling of the contrast checker message. As part of the changes, a warning icon is added and the colour of the icon/text is updated to yellow.

As part of these changes, the class names of each element has been updated to match with the BEM methodology, as per core standards.

SiobhyB committed:
Remove redundant code

'separatorType' is currently not doing anything and was added by mistake as part of the last commit, it's removed with this commit.

SiobhyB committed:
Remove border from top of contrast checker

By placing the contrast checker in its own PanelBody, we remove its top border, following design feedback.

SiobhyB commented:
@geriux, @iamthomasbishop, thank you for your reviews so far! 🙇‍♀️ I've gone ahead to implement the suggested changes to the code and have also updated the design as follows:  

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/146981246-17db7cd7-e4f0-41aa-b3b7-29afb5eed034.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/146981429-fd1f8725-a78d-4298-b4bf-4ec48663c016.png" width="100%"> |

Note, I took the values for `yellow-50` and `yellow-30` from [this color-studio list](https://github.com/Automattic/color-studio/blob/HEAD/dist/color-properties.css) that I found and added them under the `color-studio` header in [the `_colors.native.scss` file](https://github.com/WordPress/gutenberg/blob/53a1f4f06ae7e8f374217e35f40f5bb1a726a3d6/packages/base-styles/_colors.native.scss).

Let me know how this looks to you and if you have any other suggested changes.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/color-contrast-checker

geriux commented:
Hey @SiobhyB 👋 

Thanks for the changes! The code looks good! I tested it on both platforms and it works as expected 🎉 . I did notice on Android that the icon of the warning message is too close to the left margin:

<kbd><img src="https://user-images.githubusercontent.com/4885740/147084399-093ac1bd-e847-4615-8e77-bb77c3cfcba6.png" width=200 /></kbd>

SiobhyB committed:
Fix padding issue on Android

SiobhyB commented:
Ah, I slipped up on double-checking Android, thanks for spotting that @geriux! I've gone ahead to update the CSS and here's how things look across both platforms now:

<hr />

## Android

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/147135965-26a306d3-0b3c-4bbe-a1a8-0af09c503901.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/147135916-a9157954-58c4-443c-bbb7-d7421d86347c.png" width="100%"> |

| Landscape  | 
| ------------- | 
| <img src="https://user-images.githubusercontent.com/2998162/147135892-554054d1-e273-425e-9129-af6e3ae05e77.png" width="100%"> | 

## iOS

| Light mode  | Dark mode |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/147137153-7e7f3f9f-4f56-420e-bea7-c5cf109add1f.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/147136938-bd9d455b-22ef-49ad-8794-0bfa97a07a33.png" width="100%"> |

| Landscape  | 
| ------------- | 
| <img src="https://user-images.githubusercontent.com/2998162/147136996-6ed9b845-8af0-4229-8d0d-8b0c446166d3.png" width="100%"> | 

<hr />

I'm about to head off AFK for a week but looking forward to continuing to work on this in the New Year, thanks for all your help so far 🙇‍♀️

geriux reviewed (approved):
Great work! 🎉 👏  I tested this on both iOS and Android and its working as expected 🚢

SiobhyB commented:
@geriux, thanks so much! 🙇‍♀️ 

@iamthomasbishop, I'll wait to hear back on whether the updated designs match up to what you were wanting before shipping :D

iamthomasbishop commented:
@SiobhyB changes look great, thank you! Feel free to ship it when ready! 👏

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/color-contrast-checker

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/color-contrast-checker

SiobhyB committed:
Update CHANGELOG
